README.md
# Solr in Docker

## Images created

* [`alfresco-solr1`] = solr1 images, using tomcat
* [`alfresco-solr4`] = solr4 images, using tomcat
* [`alfresco-solr6`] = solr6 images (alfresco search services), using jetty

## Supported Tags

* [`:4.2`, `:4.2.8`,`:5.1`, `:5.1.4.1`] = minor, major, revision for solr1 and solr4
* [`:1.0.0`, `:1.1.1`, `:1.2.0`] = version of the alfresco search services used for solr6

## Overview

This is the repository building Solr Docker images. 

Images are automatically build by [jenkins-2](https://jenkins-2.xenit.eu) and published to [hub.xenit.eu](https://hub.xenit.eu).

## Example docker-compose files
Each time images are built, tests are automatically ran on containers started via docker-compose files.

If sharding is involved, then a valid license file which allows clustering is needed.

## Environment variables

There are several environment variables available to tweak the behaviour. The variables are read by an init script which further replaces them in the relevant files. Such relevant files include:

* solrcore.properties
* schema.xml
* server.xml (ports, ssl-related properties for tomcat-based images)
* solr.in.sh (ports, ssl-related properties, paths for jetty-based images)
* setenv.sh (JAVA_OPTS parameters for tomcat-based images)


solrcore.properties can be set via a generic mechanism by setting environment variables of the form 
GLOBAL_WORKSPACE\_<parameter> for workspace store, 
GLOBAL_ARCHIVE\_<parameter> for archive store, 
GLOBAL\_<parameter> for all cores. 

They can also be set via environment variables of the form JAVA_OPTS\_<ignored_key> where the value should be "-Dkey=value".

A subset of the properties have also dedicated environment variables e.g. ALFRESCO_ENABLE_TRACKING. Generic variables take precedence.

A subset of java variables have also dedicated environment variables e.g. JAVA_XMX. Generic variables take precedence.



Environment variables:


| Variable                    | solrcore.property variable | java variable                                                | Default                                                      | Comments |
| --------------------------- | --------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | --------------------------- |
| SHARDING                    |                      |                                                              | false                                               | if true, configuration folders for shards will be created |
| NUM_SHARDS                  | shard.count / acl.shard.count |                                                              | 3 | solr6 / solr4 |
| NUM_NODES           |                  |                                                              | 2                                                        |  |
| NODE_INSTANCE                  |                         |                                                              | 1                                                   |  |
| TEMPLATE                  | alfresco.template |                                                              | rerank                                                   |  |
| REPLICATION_FACTOR              |                     |                                                              | 1                                                        |  |
| SHARD_IDS                   | shard.instance / acl.shard.instance |                                                              | 0,1                                     | loop over values to create config folders<br />solr6 / solr4 |
| SHARD_METHOD | shard.method | | DB_ID | solr6 only |
| SHARD_KEY | shard.key | | cm:creator | solr6 only |
| CORES_TO_TRACK | | | alfresco;archive | loop over values to create config folders<br />solr6 only |
| SOLR_DATA_DIR | | | /opt/alfresco-search-services/data/index | solr6 only |
| SOLR_MODEL_DIR | | | /opt/alfresco-search-services/data/model | solr6 only |
| SOLR_CONTENT_DIR | | | /opt/alfresco-search-services/data/contentstore | solr6 only |
| ALFRESCO_SOLR_SUGGESTER_ENABLED                     | solr.suggester.enabled |                                                              | true                                                | needs also changes to schema.xml, otherwise not correct. |
| ALFRESCO_SOLR_FACETABLE_CATEGORIES_ENABLED                     |                            |                                                              |                                                          | changes schema.xml |
| ALFRESCO_HOST        | alfresco.host              |                                                              | localhost |  |
| ALFRESCO_PORT    | alfresco.port          |                                                              | 8080 |  |
| ALFRESCO_PORT_SSL | alfresco.port.ssl      |                                                              | 8443 |  |
| ALFRESCO_SSL | alfresco.secureComms | | https | changes also server.xml for tomcat-based images and solr.in.sh for jetty-based images |
| ALFRESCO_ENABLE_TRACKING | enable.alfresco.tracking    |                                                              | true | in the workspace core or shard |
| ALFRESCO_INDEX_CONTENT | alfresco.index.transformContent |                                                              | true | in the workspace core or shard |
| ALFRESCO_CORE_POOL_SIZE | alfresco.corePoolSize |  | 8 |  |
| ALFRESCO_DO_PERMISSION_CHECKS | alfresco.doPermissionChecks |  | true | post filtering of results still happens on Alfresco side |
| ARCHIVE_ENABLE_TRACKING | enable.alfresco.tracking |  | true | in the archive core |
| ARCHIVE_INDEX_CONTENT | alfresco.index.transformContent |                                                              | true | in the archive core |
| TOMCAT_PORT |  | -DTOMCAT_PORT | 8080 | solr1 and solr4 only |
| TOMCAT_PORT_SSL | | -DTOMCAT_PORT_SSL | 8443 | solr1 and solr4 only |
| TOMCAT_AJP_PORT | | -DTOMCAT_AJP_PORT | 8009 | solr1 and solr4 only |
| TOMCAT_SERVER_PORT | | -DTOMCAT_SERVER_PORT | 8005 | solr1 and solr4 only |
| TOMCAT_MAX_HTTP_HEADER_SIZE | | -DTOMCAT_MAX_HTTP_HEADER_SIZE  or -DMAX_HTTP_HEADER_SIZE | 32768 | solr1 and solr4 only |
| TOMCAT_MAX_THREADS | | -DTOMCAT_MAX_THREADS or -DMAX_THREADS | 200 | solr1 and solr4 only |
| JETTY_PORT | |  | 8080 | solr6 only |
| JETTY_PORT_SSL | |  | 8443 | solr6 only |
| JAVA_XMX | | -Xmx | 2048M | |
| JAVA_XMS | | -Xms | 512M | |
| JMX_ENABLED | | -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.local.only=false -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.rmi.port=5000 -Dcom.sun.management.jmxremote.port=5000 -Djava.rmi.server.hostname=$JMX_RMI_HOST | false | |
| JMX_RMI_HOST | |  | 0.0.0.0 | |
| DEBUG | | -Xdebug -Xrunjdwp:transport=dt_socket, address=8000, server=y, suspend=n | false | |
| GLOBAL_WORKSPACE_<variable> | <variable> | | | for workspace core or shards |
| GLOBAL_ARCHIVE_<variable> | <variable> | | | for archive core |
| GLOBAL_<variable> | <variable> | | |  |
| JAVA_OPTS_<ignored_variable>=<value> |  | <value> | | |


## Docker-compose files

Besides the docker-compose files used in the tests, there are other example files in [src/main/resources](https://bitbucket.org/xenit/docker-solr/src/master/src/main/resources/).

## Support & Collaboration

Issues can be reported on the [JIRA DOCKER](https://xenitsupport.jira.com/projects/DOCKER) project.

These images are updated via pull requests to the [xenit/docker-solr/](https://bitbucket.org/xenit/docker-solr/) BitBucket-repository.


## FAQ

### How do I access the Tomcat debugport ?

Set the environment variable DEBUG=true. The debug port is 8000.

### How do I enable JMX?

Set the environment variable JMX_ENABLED=true. Jmx port is 5000.




Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [v0.0.5] - 2018-10-26
### Added
* [DOCKER-112] Integration with vault

### Fixed
* [DOCKER-135] Remove file appender, the only logs are on stdout and stderr which can be externalized

## [v0.0.4] - 2018-09-25
### Added
* [DOCKER-115] Move healthchecks to Dockerfiles
* [DOCKER-119] Generic mechanism to set variables via JAVA_OPTS_<variable>
* [DOCKER-121] Options for suggester and facetable categories for solr6
* [DOCKER-118] Generic mechanism to set variables via GLOBAL_WORKSPACE_, GLOBAL_ARCHIVE_, GLOBAL_ environment variables

### Fixed
* [DOCKER-91] Redirect port was not parametrized in server.xml, therefore it was not possible to change it
* [DOCKER-116] Change ports via variables JETTY_PORT and JETTY_PORT_SSL
* [DOCKER-122] Solr1, solr4 and solr6 run with the same uid+gid

### Changed
* [DOCKER-109] Adapt all solr images to start from Xenit base images

### Removed
* [DOCKER-120] Removed variables SOLR_PORT, SOLR_PORT_SSL = they are set with TOMCAT_PORT, TOMCAT_PORT_SSL and JETTY_PORT, JETTY_PORT_SSL
* [DOCKER-117] Removed custom variables for caches. They can be set via GLOBAL_WORKSPACE_, GLOBAL_ARCHIVE_, GLOBAL_ options

## [v0.0.3] - 2018-09-10
### Added
* [DOCKER-77] Smoke tests: search + status for shards (if testsSharded = true)
* [DOCKER-104] Alfresco search services 1.2.0

### Changed
* [DOCKER-90] Restructuring: global + local resources, single build
 
## [v0.0.2] - 2018-09-04
### Added
* [DOCKER-106] Parameter CORES_TO_TRACK as a ; separated list: e.g. CORES_TO_TRACK=alfresco;archive;version 
* [DOCKER-103] Parameters SOLR_DATA_DIR, SOLR_MODEL_DIR, SOLR_CONTENT_DIR to changed default locations
* [DOCKER-88] Solr1
* [DOCKER-56] Alfresco search services 1.1.1

### Changed
* [DOCKER-105] Refactored core creation: use static creation for default cores, same mechanism as for shards
* [DOCKER-66] Naming images 	
	
## [v0.0.1] - 2018-05-23
### Added
* [DOCKER-39] Support for JAVA_OPTS_<variable> variables, allowing for overrides in different docker-compose files.
* [DOCKER-41] Support for SSL in Alfresco search services 1.0.0 (solr6)



src/test/java/eu/xenit/docker/solr/test/SolrSmokeTests.java
package eu.xenit.docker.solr.test;

import io.restassured.RestAssured;
import io.restassured.authentication.PreemptiveBasicAuthScheme;
import io.restassured.builder.RequestSpecBuilder;
import io.restassured.parsing.Parser;
import io.restassured.specification.RequestSpecification;
import org.junit.BeforeClass;
import org.junit.Test;

import static io.restassured.RestAssured.given;
import static io.restassured.http.ContentType.JSON;
import static java.lang.Thread.sleep;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.equalToIgnoringCase;
import static org.hamcrest.Matchers.greaterThan;
import static org.junit.Assert.*;

public class SolrSmokeTests {
    static RequestSpecification spec;
    static RequestSpecification specSharded;
    static RequestSpecification specShardedSolr1;
    static RequestSpecification specShardedSolr2;

    @BeforeClass
    public static void setup() {

        String basePath = "/alfresco/s/";
        String basePathSolr = "solr/admin/cores";
        if("solr4".equals(System.getProperty("flavor")))
            basePathSolr = "solr4/admin/cores";
        String host = System.getProperty("alfresco.host");
        String hostSharded = System.getProperty("alfresco-sharded.host");
        String solr1 = System.getProperty("solr1-sharded.host");
        String solr2 = System.getProperty("solr2-sharded.host");
        int port = Integer.parseInt(System.getProperty("alfresco.tcp.8080"));
        int portSharded = ((hostSharded!=null)?Integer.parseInt(System.getProperty("alfresco-sharded.tcp.8080")):-1);
        int portShardedSolr1 = ((hostSharded!=null)?Integer.parseInt(System.getProperty("solr1-sharded.tcp.8080")):-1);
        int portShardedSolr2 = ((hostSharded!=null)?Integer.parseInt(System.getProperty("solr2-sharded.tcp.8080")):-1);

        String baseURI = "http://" + host;
        String baseURISharded = "http://" + hostSharded;
        String baseURIShardedSolr1 = "http://" + solr1;
        String baseURIShardedSolr2 = "http://" + solr2;

        PreemptiveBasicAuthScheme authScheme = new PreemptiveBasicAuthScheme();
        authScheme.setUserName("admin");
        authScheme.setPassword("admin");
        RestAssured.defaultParser = Parser.JSON;

        spec = new RequestSpecBuilder()
                .setBaseUri(baseURI)
                .setPort(port)
                .setBasePath(basePath)
                .setAuth(authScheme)
                .build();
        if(hostSharded != null) {
            specSharded = new RequestSpecBuilder()
                    .setBaseUri(baseURISharded)
                    .setPort(portSharded)
                    .setBasePath(basePath)
                    .setAuth(authScheme)
                    .build();
            specShardedSolr1 = new RequestSpecBuilder()
                    .setBaseUri(baseURIShardedSolr1)
                    .setPort(portShardedSolr1)
                    .setBasePath(basePathSolr)
                    .addParam("action","STATUS")
                    .addParam("wt","json")
                    .build();
            specShardedSolr2 = new RequestSpecBuilder()
                    .setBaseUri(baseURIShardedSolr2)
                    .setPort(portShardedSolr2)
                    .setBasePath(basePathSolr)
                    .addParam("action","STATUS")
                    .addParam("wt","json")
                    .build();
        } else {
            specSharded = null;
            specShardedSolr1 = null;
            specShardedSolr2 = null;
        }
    }


    @Test
    // Note: due to eventual consistency, we can't test the expected number of results
    public void testSearch(){
        given()
                .spec(spec)
                .when()
                .get("slingshot/search?term=test*")
                .then()
                .statusCode(200);
        if(specSharded!=null)
            given()
                    .spec(specSharded)
                    .when()
                    .get("slingshot/search?term=test*")
                    .then()
                    .statusCode(200);
    }

    @Test
    public void TestShards() {
        if(specSharded!=null) {
            try {
                // wait for solr to track
                sleep(30000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        if(specShardedSolr1!=null) {
            Integer docs0 = given()
                    .spec(specShardedSolr1)
                    .contentType("application/json")
                    .when()
                    .get()
                    .then()
                    .statusCode(200)
                    .contentType(JSON)
                    .extract().path("status.alfresco-0.index.numDocs");
            Integer docs1 = given()
                    .spec(specShardedSolr1)
                    .contentType("application/json")
                    .when()
                    .get()
                    .then()
                    .statusCode(200)
                    .contentType(JSON)
                    .extract().path("status.alfresco-1.index.numDocs");
            Integer docs2 = given()
                    .spec(specShardedSolr2)
                    .contentType("application/json")
                    .when()
                    .get()
                    .then()
                    .statusCode(200)
                    .contentType(JSON)
                    .extract().path("status.alfresco-2.index.numDocs");
            assertThat(docs0,greaterThan(50));
            assertThat(docs1,greaterThan(50));
            assertThat(docs2,greaterThan(50));
        }
    }
}


solr6/local/init.sh
#!/bin/bash

# should get environment var from docker image / container:
# ${SOLR_INSTALL_HOME} = /opt/alfresco-search-services

set -e

SOLR_DIR_ROOT='/opt/alfresco-search-services/solrhome'
SOLR_HOST='localhost'

DEBUG=${DEBUG:-'false'}
ALFRESCO_SSL=${ALFRESCO_SSL:-'https'}
JETTY_PORT=${JETTY_PORT:-'8080'}
JETTY_PORT_SSL=${JETTY_PORT_SSL:-'8443'}
JMX_ENABLED=${JMX_ENABLED:-'false'}
JMX_RMI_HOST=${JMX_RMI_HOST:-'0.0.0.0'}
CORES_TO_TRACK=${CORES_TO_TRACK:-"alfresco;archive"}
IFS=';' read -r -a DEFAULT_CORES <<< "$CORES_TO_TRACK"

if [ $ALFRESCO_SSL != none ]
then
    PORT=$JETTY_PORT_SSL
else
    PORT=$JETTY_PORT
fi    

SHARDING=${SHARDING:-'false'}
NUM_SHARDS=${NUM_SHARDS:-'3'}
NUM_NODES=${NUM_NODES:-'2'}
NODE_INSTANCE=${NODE_INSTANCE:-'1'}
TEMPLATE=${TEMPLATE:-'rerank'}
REPLICATION_FACTOR=${REPLICATION_FACTOR:-'1'}
SHARD_IDS=${SHARD_IDS:-'0,1'}

ALFRESCO_SOLR_SUGGESTER_ENABLED=${ALFRESCO_SOLR_SUGGESTER_ENABLED:-'true'}
ALFRESCO_SOLR_FACETABLE_CATEGORIES_ENABLED=${ALFRESCO_SOLR_FACETABLE_CATEGORIES_ENABLED:-'false'}

JAVA_XMS=${JAVA_XMS:-'512M'}
JAVA_XMX=${JAVA_XMX:-'2048M'}

CONFIG_FILE_SOLR_START="$SOLR_INSTALL_HOME/solr.in.sh"

sed -i 's/log4j.rootLogger=INFO, file, CONSOLE/log4j.rootLogger=ERROR, CONSOLE/' /opt/alfresco-search-services/logs/log4j.properties

function setJavaOption {
    JAVA_OPTS="$JAVA_OPTS $2"
}

function setJavaOptions {
    IFS=$'\n'
    for i in `env`
    do
	if [[ $i == JAVA_OPTS_* ]]
	    then
	    key=`echo $i | cut -d '=' -f 1 | cut -d '_' -f 3-`
	    value=`echo $i | cut -d '=' -f 2-`
	    setJavaOption $key $value
	fi
    done
}

function setOption {
    if grep --quiet -e "$1\s*=" "$3"; then
        # replace option
        sed -i "s#^\($1\s*=\s*\).*\$#\1$2#" $3
	    sed -i "s#^\#\($1\s*=\s*\).*\$#\1$2#" $3
        if (( $? )); then
            echo "setOption failed (replacing option $1=$2 in $3)"
            exit 1
        fi
    else
        # add option if it does not exist
        echo "$1=$2" >> $3
    fi
}

function setGlobalOptions {
    file=$1
    coreName=$2
    IFS=$'\n'
    for i in `env`
    do
	if [[ $i == GLOBAL_WORKSPACE_* ]]
	then
	    if [ $coreName = alfresco ]
	    then
	        key=`echo $i | cut -d '=' -f 1 | cut -d '_' -f 3-`
	        value=`echo $i | cut -d '=' -f 2-`
	        setOption $key $value "$file"
	    fi
	elif [[ $i == GLOBAL_ARCHIVE_* ]]
	then
	    if [ $coreName = archive ]
	    then
	        key=`echo $i | cut -d '=' -f 1 | cut -d '_' -f 3-`
	        value=`echo $i | cut -d '=' -f 2-`
	        setOption $key $value "$file"
	    fi
	elif [[ $i == GLOBAL_* ]]
	then
            key=`echo $i | cut -d '=' -f 1 | cut -d '_' -f 2-`
            value=`echo $i | cut -d '=' -f 2-`
            setOption $key $value "$file"
	fi
    done
}

function escapeFile {
    sed -i "s#:#\:#g" $1
}

function createCoreStatically {
    coreName="$1"
    solrCoreName="$2"
    newCore="$3"

    echo "Creating Alfresco core=$coreName, solrCore=$solrCoreName in $newCore"
    CONFIG_FILE_CORE=$newCore/conf/solrcore.properties
    if [ ! -f  "$newCore" ]
    then
	cp -r ${SOLR_DIR_ROOT}/templates/$TEMPLATE $newCore
	FILE_CORE=$newCore/core.properties
	touch $FILE_CORE
	setOption 'name' "$solrCoreName" "$FILE_CORE"	
    fi
    setOption 'data.dir.root' "${SOLR_DATA_DIR:-/opt/alfresco-search-services/data/index}" "$CONFIG_FILE_CORE"
    setOption 'data.dir.store' "$coreName" "$CONFIG_FILE_CORE"
    setOption 'alfresco.template' "${TEMPLATE}" "$CONFIG_FILE_CORE"
    setOption 'alfresco.host' "${ALFRESCO_HOST:-localhost}" "$CONFIG_FILE_CORE"
    setOption 'alfresco.port' "${ALFRESCO_PORT:-8080}" "$CONFIG_FILE_CORE"
    setOption 'alfresco.port.ssl' "${ALFRESCO_PORT_SSL:-8443}" "$CONFIG_FILE_CORE"
    setOption 'alfresco.secureComms' "$ALFRESCO_SSL" "$CONFIG_FILE_CORE"
}

function makeConfigs {

    for coreName in "${DEFAULT_CORES[@]}"
    do
	newCore=${SOLR_DIR_ROOT}/$coreName
	CONFIG_FILE_CORE=$newCore/conf/solrcore.properties
	
	if [ $coreName = alfresco ]
	then
	    # for sharding - dynamic registration of shards
	    if [ $SHARDING = true ]
	    then
		# assuming we shard the workspace store
		rm -rf ${SOLR_DIR_ROOT}/workspace-SpacesStore
		collectionName=${TEMPLATE}--workspace-SpacesStore--shards--$NUM_SHARDS-x-$REPLICATION_FACTOR--node--$NODE_INSTANCE-of-$NUM_NODES
		echo "collectionName=$collectionName for SHARD_IDS=$SHARD_IDS"
		mkdir -p ${SOLR_DIR_ROOT}/$collectionName
		for i in $(echo $SHARD_IDS | tr "," "\n")
		do
		    coreName=workspace-SpacesStore-$i
		    solrCoreName=alfresco-$i
		    SHARED_PROPERTIES=${SOLR_DIR_ROOT}/conf/shared.properties
		    newCore=${SOLR_DIR_ROOT}/$collectionName/$coreName
		    CONFIG_FILE_CORE=$newCore/conf/solrcore.properties
		    
		    createCoreStatically "$coreName" "$solrCoreName" "$newCore"
		    
		    setOption 'enable.alfresco.tracking' "${ALFRESCO_ENABLE_TRACKING:-true}" "$CONFIG_FILE_CORE"
		    setOption 'alfresco.index.transformContent' "${ALFRESCO_INDEX_CONTENT:-true}" "$CONFIG_FILE_CORE"
		    setOption 'alfresco.corePoolSize' "${ALFRESCO_CORE_POOL_SIZE:-8}" "$CONFIG_FILE_CORE"
		    setOption 'alfresco.doPermissionChecks' "${ALFRESCO_DO_PERMISSION_CHECKS:-true}" "$CONFIG_FILE_CORE"
		    setOption 'solr.suggester.enabled' "$ALFRESCO_SOLR_SUGGESTER_ENABLED" "$CONFIG_FILE_CORE"
		    setOption 'shard.method' "${SHARD_METHOD:-DB_ID}" "$CONFIG_FILE_CORE"
		    setOption 'shard.key' "${SHARD_KEY:-cm:creator}" "$CONFIG_FILE_CORE"
		    setOption 'shard.count' "$NUM_SHARDS" "$CONFIG_FILE_CORE"
		    setOption 'shard.instance' "$i" "$CONFIG_FILE_CORE"         

		    CONFIG_FILE_SOLR_SCHEMA=$newCore/conf/schema.xml
		    if [ $ALFRESCO_SOLR_SUGGESTER_ENABLED = true ]
		    then
		    	sed -i 's/.*\(<copyField source="suggest_\*" dest="suggest" \/>\).*/\1/g' "$CONFIG_FILE_SOLR_SCHEMA"
		    else
			    sed -i 's/.*\(<copyField source="suggest_\*" dest="suggest" \/>\).*/<!--\1-->/g' "$CONFIG_FILE_SOLR_SCHEMA"
		    fi
		    if [ $ALFRESCO_SOLR_FACETABLE_CATEGORIES_ENABLED = true ]
		    then
			    sed -i 's/\(.*<dynamicField.*name="\(category\|noderef\)@m_.*type="\)\(oldStandardAnalysis\)\(".*\)\(\/\)\(.*\)/\1identifier\4docValues="true" \/\6/g' "$CONFIG_FILE_SOLR_SCHEMA"
			    sed -i 's/\(.*<dynamicField.*name="\(category\|noderef\)@s_.*type="\)\(oldStandardAnalysis\)\(".*\)\(sortMissingLast="true"\)\(.*\)/\1identifier\4docValues="true"\6/g' "$CONFIG_FILE_SOLR_SCHEMA"
		    fi
		    
		    setGlobalOptions "$CONFIG_FILE_CORE" alfresco
		    
		    setOption 'solr.host' "$SOLR_HOST" "$SHARED_PROPERTIES"
		    setOption 'solr.port' "$PORT" "$SHARED_PROPERTIES"
		    
		    escapeFile "$CONFIG_FILE_CORE"
		done
	    else
		createCoreStatically "alfresco" "alfresco" "$newCore"
		
		setOption 'alfresco.stores' "workspace://SpacesStore" "$CONFIG_FILE_CORE"
		setOption 'enable.alfresco.tracking' "${ALFRESCO_ENABLE_TRACKING:-true}" "$CONFIG_FILE_CORE"
		setOption 'alfresco.index.transformContent' "${ALFRESCO_INDEX_CONTENT:-true}" "$CONFIG_FILE_CORE"
		setOption 'alfresco.corePoolSize' "${ALFRESCO_CORE_POOL_SIZE:-8}" "$CONFIG_FILE_CORE"
		setOption 'alfresco.doPermissionChecks' "${ALFRESCO_DO_PERMISSION_CHECKS:-true}" "$CONFIG_FILE_CORE"
		setOption 'solr.suggester.enabled' "$ALFRESCO_SOLR_SUGGESTER_ENABLED" "$CONFIG_FILE_CORE"

		CONFIG_FILE_SOLR_SCHEMA=$newCore/conf/schema.xml
		if [ $ALFRESCO_SOLR_SUGGESTER_ENABLED = true ]
		then
		    sed -i 's/.*\(<copyField source="suggest_\*" dest="suggest" \/>\).*/\1/g' "$CONFIG_FILE_SOLR_SCHEMA"
		else
			sed -i 's/.*\(<copyField source="suggest_\*" dest="suggest" \/>\).*/<!--\1-->/g' "$CONFIG_FILE_SOLR_SCHEMA"
		fi
		if [ $ALFRESCO_SOLR_FACETABLE_CATEGORIES_ENABLED = true ]
		then
			sed -i 's/\(.*<dynamicField.*name="\(category\|noderef\)@m_.*type="\)\(oldStandardAnalysis\)\(".*\)\(\/\)\(.*\)/\1identifier\4docValues="true" \/\6/g' "$CONFIG_FILE_SOLR_SCHEMA"
			sed -i 's/\(.*<dynamicField.*name="\(category\|noderef\)@s_.*type="\)\(oldStandardAnalysis\)\(".*\)\(sortMissingLast="true"\)\(.*\)/\1identifier\4docValues="true"\6/g' "$CONFIG_FILE_SOLR_SCHEMA"
		fi

		    setGlobalOptions "$CONFIG_FILE_CORE" alfresco
		    escapeFile "$CONFIG_FILE_CORE"
		fi
		elif [ $coreName = archive ]
		then
		    createCoreStatically "archive" "archive" "$newCore"
		    
		    setOption 'alfresco.stores' "archive://SpacesStore" "$CONFIG_FILE_CORE"
		    setOption 'enable.alfresco.tracking' "${ARCHIVE_ENABLE_TRACKING:-true}" "$CONFIG_FILE_CORE"
		    setOption 'alfresco.index.transformContent' "${ARCHIVE_INDEX_CONTENT:-true}" "$CONFIG_FILE_CORE"

		    setGlobalOptions "$CONFIG_FILE_CORE" archive
		    escapeFile "$CONFIG_FILE_CORE"	
		    elif [ $coreName = version ]
		    then
		    createCoreStatically "version" "version" "$newCore"
		    
		    setOption 'alfresco.stores' "workspace://version2Store" "$CONFIG_FILE_CORE"

		    setGlobalOptions "$CONFIG_FILE_CORE" version
		    escapeFile "$CONFIG_FILE_CORE"
		    else
			"Core $coreName not found"
		fi
		done
		
		if [ $JMX_ENABLED = true ]
		then
		    JAVA_OPTS="${JAVA_OPTS} -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.local.only=false -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.rmi.port=5001 -Dcom.sun.management.jmxremote.port=5001 -Djava.rmi.server.hostname=${JMX_RMI_HOST}"
		fi
		if [ $DEBUG = true ]
		then
		    JAVA_OPTS="${JAVA_OPTS} -Xdebug -Xrunjdwp:transport=dt_socket,address=8000,server=y,suspend=n"
		fi
}

#follow http://docs.alfresco.com/5.2/tasks/solr6-install.html
echo "export SOLR_SOLR_DATA_DIR=${SOLR_DATA_DIR:-/opt/alfresco-search-services/data/index}" >>"$CONFIG_FILE_SOLR_START"
echo "export SOLR_SOLR_MODEL_DIR=${SOLR_MODEL_DIR:-/opt/alfresco-search-services/data/model}" >>"$CONFIG_FILE_SOLR_START"
echo "export SOLR_SOLR_CONTENT_DIR=${SOLR_CONTENT_DIR:-/opt/alfresco-search-services/data/contentstore}" >>"$CONFIG_FILE_SOLR_START"

if [ $ALFRESCO_SSL != none ]
then
    cp -r ${SOLR_DIR_ROOT}/keystore/* ${SOLR_DIR_ROOT}/templates/${TEMPLATE}/conf/
    
    setOption 'SOLR_SSL_KEY_STORE' "${SOLR_DIR_ROOT}/keystore/ssl.repo.client.keystore" "$CONFIG_FILE_SOLR_START"
    setOption 'SOLR_SSL_KEY_STORE_PASSWORD' "kT9X6oe68t" "$CONFIG_FILE_SOLR_START"
    setOption 'SOLR_SSL_TRUST_STORE' "${SOLR_DIR_ROOT}/keystore/ssl.repo.client.truststore" "$CONFIG_FILE_SOLR_START"
    setOption 'SOLR_SSL_TRUST_STORE_PASSWORD' "kT9X6oe68t" "$CONFIG_FILE_SOLR_START"
    setOption 'SOLR_SSL_NEED_CLIENT_AUTH' "true" "$CONFIG_FILE_SOLR_START"
    setOption 'SOLR_SSL_WANT_CLIENT_AUTH' "false" "$CONFIG_FILE_SOLR_START"
    setOption 'SOLR_SOLR_HOST' "$SOLR_HOST" "$CONFIG_FILE_SOLR_START"
    setOption 'SOLR_SOLR_PORT' "$PORT" "$CONFIG_FILE_SOLR_START"
    
    setJavaOption 'javax.net.ssl.keyStoreType' '-Djavax.net.ssl.keyStoreType=JCEKS'
    setJavaOption 'javax.net.ssl.trustStoreType' '-Djavax.net.ssl.trustStoreType=JCEKS'
    setJavaOption 'solr.ssl.checkPeerName' '-Dsolr.ssl.checkPeerName=false'        
fi



appStart () {
    mkdir -p /opt/alfresco-search-services/data/
    setJavaOptions
    makeConfigs
    echo "${SOLR_INSTALL_HOME}/solr/bin/solr start -f -m ${JAVA_XMX} -p ${PORT} -h ${SOLR_HOST} -s ${SOLR_DIR_ROOT} -a ${JAVA_OPTS}"
    ${SOLR_INSTALL_HOME}/solr/bin/solr start -f -m ${JAVA_XMX} -p ${PORT} -h ${SOLR_HOST} -s ${SOLR_DIR_ROOT} -a "${JAVA_OPTS}"
}

appHelp () {
    echo "Available options:"
    echo " app:start    - Start the application"
    echo " app:help     - Displays the help"
    echo " [command]    - Execute the specified linux command eg. bash."
}

case "$1" in
    app:start)
	appStart
	;;
    app:help)
	appHelp
	;;
    *)
	if [ -x $1 ]; then
	    $1
	else
	    prog=$(which $1)
	    if [ -n "${prog}" ] ; then
		shift 1
		$prog $@
	    else
		appHelp
	    fi
	fi
	;;
esac

exit 0
